src/main/java/graphql/normalized/ExecutableNormalizedOperationToAstCompiler.java

```diff
@@ -26,71 +29,81 @@
+                                             List<ExecutableNormalizedField> topLevelFields) {
+        List<Selection<?>> selections = selectionsForNormalizedFields(schema, topLevelFields);
         SelectionSet selectionSet = new SelectionSet(selections);
 
         return Document.newDocument()
-                .definition(OperationDefinition.newOperationDefinition()
-                        .name(operationName)
-                        .operation(operationKind)
-                        .selectionSet(selectionSet)
-                        .build())
-                .build();
+            .definition(OperationDefinition.newOperationDefinition()
+                .name(operationName)
+                .operation(operationKind)
+                .selectionSet(selectionSet)
+                .build())
+            .build();
     }
 
-    private static List<Selection<?>> selectionsForNormalizedFields(List<ExecutableNormalizedField> executableNormalizedFields) {
-        ImmutableList.Builder<Selection<?>> result = ImmutableList.builder();
-
-        Map<String, List<Field>> overallGroupedFields = new LinkedHashMap<>();
-        for (ExecutableNormalizedField nf : executableNormalizedFields) {
-            Map<String, List<Field>> groupFieldsForChild = selectionForNormalizedField(nf);
+    private static List<Selection<?>> selectionsForNormalizedFields(GraphQLSchema schema,
+                                                                    List<ExecutableNormalizedField> executableNormalizedFields) {
+        ImmutableList.Builder<Selection<?>> selections = ImmutableList.builder();
 
-            groupFieldsForChild.forEach((objectTypeName, fields) -> {
-                List<Field> fieldList = overallGroupedFields.computeIfAbsent(objectTypeName, ignored -> new ArrayList<>());
-                fieldList.addAll(fields);
-            });
+        // All conditional fields go here instead of directly to selections so they can be grouped together
+        // in the same inline fragement in the output
+        Map<String, List<Field>> conditionalFieldsByObjectTypeName = new LinkedHashMap<>();
 
+        for (ExecutableNormalizedField nf : executableNormalizedFields) {
+            Map<String, List<Field>> groupFieldsForChild = selectionForNormalizedField(schema, nf);
+            if (nf.isConditional(schema)) {
+                groupFieldsForChild.forEach((objectTypeName, fields) -> {
+                    List<Field> fieldList = conditionalFieldsByObjectTypeName.computeIfAbsent(objectTypeName, ignored -> new ArrayList<>());
+                    fieldList.addAll(fields);
+                });
+            } else {
+                List<Field> fields = groupFieldsForChild.values().iterator().next();
+                selections.addAll(fields);
+            }
         }
 
-        overallGroupedFields.forEach((objectTypeName, fields) -> {
+        conditionalFieldsByObjectTypeName.forEach((objectTypeName, fields) -> {
```

Comment:
As usual, we group the condition fields by object type conditions so that they can be grouped together in the same inline fragment.